### PR TITLE
Support confirm signin to pass user attribute

### PR DESCRIPTION
**Pull request by royjit**
*Issue #1248*

*Description of changes:*

Confirm attribute require user attribute while changing password for FORCE_CHANGE_PASSWORD state.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClientExtensions.swift [hunk 1]
```diff
@@ -640,14 +640,17 @@ extension AWSMobileClient {
     ///
     /// - Parameters:
     ///   - challengeResponse: confirmation code or TOTP token which is available to the user.
+    ///   - userAttributes: user attributes required for the operation.
     ///   - completionHandler: completionHandler which will be called when result is available.
-    public func confirmSignIn(challengeResponse: String, completionHandler: @escaping ((SignInResult?, Error?) -> Void)) {
+    public func confirmSignIn(challengeResponse: String,
+                              userAttributes:[String:String] = [:],
```
Comment (palpatim) [inline AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClientExtensions.swift]: nit: spacing :)

File: AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClientExtensions.swift [hunk 2]
```diff
@@ -640,14 +640,17 @@ extension AWSMobileClient {
...
+                              completionHandler: @escaping ((SignInResult?, Error?) -> Void)) {
         if (self.userpoolOpsHelper.mfaCodeCompletionSource != nil) {
             self.userpoolOpsHelper.currentSignInHandlerCallback = completionHandler
             self.userpoolOpsHelper.mfaCodeCompletionSource?.set(result: challengeResponse as NSString)
         } else if (self.userpoolOpsHelper.newPasswordRequiredTaskCompletionSource != nil) {
             self.userpoolOpsHelper.currentSignInHandlerCallback = completionHandler
-            self.userpoolOpsHelper.newPasswordRequiredTaskCompletionSource?.set(result: AWSCognitoIdentityNewPasswordRequiredDetails.init(proposedPassword: challengeResponse, userAttributes: [String:String]()))
+            self.userpoolOpsHelper.newPasswordRequiredTaskCompletionSource?.set(result: AWSCognitoIdentityNewPasswordRequiredDetails.init(proposedPassword: challengeResponse, userAttributes: userAttributes))
```
Comment (palpatim) [inline AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClientExtensions.swift]: Can we add a test asserting this is passed through?

File: AWSAuthSDK/Tests/AWSMobileClientTests/AWSMobileClientTests.swift [hunk 3]
```diff
@@ -136,33 +136,67 @@ class AWSMobileClientTests: XCTestCase {
...
             }
 
             if let signInResult = signInResult {
-                switch(signInResult.signInState) {
-                case .signedIn:
-                    print("signed in")
-                default:
-                    XCTFail("User was not signed in, un-expected.")
-                }
+                XCTAssertEqual(signInResult.signInState, verifySignState, "Could not verify sign in state")
```
Comment (palpatim) [inline AWSAuthSDK/Tests/AWSMobileClientTests/AWSMobileClientTests.swift]: If we're not expecting signInResult to be `nil` in the normal flow of the test, we should just move this out and remove the `if let` block. Alternately, change it to a `guard let signInResult = signInResult` and `XCTFail("signInResult unexpectedly nil")` or similar

File: AWSAuthSDK/Tests/AWSMobileClientTests/AWSMobileClientTests.swift [hunk 4]
```diff
@@ -136,33 +136,67 @@ class AWSMobileClientTests: XCTestCase {
...
             signInWasSuccessful.fulfill()
         }
         wait(for: [signInWasSuccessful], timeout: 5)
     }
     
+    func confirmSign(challengeResponse: String, userAttributes:[String:String] = [:], verifySignState: SignInState = .signedIn) {
+        
+        let signInConfirmWasSuccessful = expectation(description: "signIn confirm was successful")
+        AWSMobileClient.sharedInstance().confirmSignIn(challengeResponse: challengeResponse,
+                                                       userAttributes: userAttributes) {
```
Comment (palpatim) [inline AWSAuthSDK/Tests/AWSMobileClientTests/AWSMobileClientTests.swift]: As noted, we should also add a test asserting the user attributes are passed through

Comment (palpatim) [reply AWSAuthSDK/Tests/AWSMobileClientTests/AWSMobileClientTests.swift]: Talked offline, this isn't needed.

File: AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClientExtensions.swift (hunk 2 above)
Comment (palpatim) [reply AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClientExtensions.swift]: Talked offline, this isn't needed.

File: AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClientExtensions.swift (hunk 1 above)
Comment (royjit) [reply AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClientExtensions.swift]: Fixed

File: AWSAuthSDK/Tests/AWSMobileClientTests/AWSMobileClientTests.swift (hunk 3 above)
Comment (royjit) [reply AWSAuthSDK/Tests/AWSMobileClientTests/AWSMobileClientTests.swift]: Added check for signInResult

File: AWSAuthSDK/Tests/AWSMobileClientTests/AWSMobileClientTests.swift [hunk 5]
```diff
@@ -136,33 +136,69 @@ class AWSMobileClientTests: XCTestCase {
...
```
Comment (kvasukib) [inline AWSAuthSDK/Tests/AWSMobileClientTests/AWSMobileClientTests.swift]: What userAttributes are being passed in here?

Comment (palpatim) [reply AWSAuthSDK/Tests/AWSMobileClientTests/AWSMobileClientTests.swift]: @kvasukib Attributes required by the User Pool config must be passed in during the password confirmation flow:

https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_AdminInitiateAuth.html

> NEW_PASSWORD_REQUIRED: For users which are required to change their passwords after successful first login. This challenge should be passed with NEW_PASSWORD and any other required attributes.

**Pull request merged**